snap-gpf/src/main/java/org/esa/snap/core/gpf/internal/TileImpl.java

```diff
@@ -551,9 +605,33 @@ public void setSample(int x, int y, double sample) {
         if (scaled) {
             sample = toRaw(sample);
         }
+        switch (rasterDataNode.getDataType()) {
```

Comment:
If raw data run into a clipping there must be something wrong with geopysical data because they run outside the covered valid value range, defined by raw data type and scaling. In this case clipping the raw data value then turns a wrong geophysical value into a valid value.
This can not be right.

This should be done that way:
First calculate the geopysical clipping bound by scaling the raw min and raw max value (take care if signed or unsigned), then clip the geopysical, then transform to raw and round to prevent the cast error.

An integer cast is neither a floor nor a round nor a cail operation.
A cast on a positive double value is equal to a floor operation.
But on negative values it is equal to a cail operation.
It changes the direction of the value change for negative numbers.

Reply:
I disagree. 

1. A geophysical value that is outside the range of the raw data type is not invalid. It is just to large to be stored. it is better to clip it than to pass it uncontrolled.
2. Casting of a `double` or `float` to a `short` or `byte` yields rather weird results. Cast of a positive `double` or `float` to a `short` or `byte` is not equal to a `floor` operation. Test it yourself.
3. A scientist developing an operator probably does not want to bother with clipping and data types. Often the data type is predetermined (from external data) and not always the same.
4. The API documentation of `setSample` states that the values are clipped, which they are not.

Reply:
> 1. A geophysical value that is outside the range of the raw data type is not invalid. It is just too large to be stored. It is better to clip it than to pass it uncontrolled.

Thats exactly what I meant.
Values are invalid from the point of view of "which data can be saved" not from the Point of view, does the value make sence from the scientific poit of view of valid data range.

If the value is in a scientific valid range, but can not be saved, then the configuration of the store must be wrong. Then either the raw data type or the scaling must be wrong.

If a value is a scientific valid value, a raw clipping changes this scientific valid value. Does this make sence?

Ideally a raw date store should not manipulate data. Also a raw data store shall not know anything of the valid scientific value range. A raw data store shall only be able to tore store all the scientivic valid data of a cerain use case. For this a raw data store has to be configured the right way. Then scientific valid data can be stored using the scaling but without manipulation.